modules/substrate/src/lib.rs

```diff
@@ -309,18 +329,44 @@ impl<T: Trait> BridgeStorage for PalletStorage<T> {
 	type Header = BridgedHeader<T>;
 
 	fn write_header(&mut self, header: &ImportedHeader<BridgedHeader<T>>) {
-		let hash = header.header.hash();
+		use core::cmp::Ordering;
+
+		let hash = header.hash();
+		let current_height = header.number();
+		let best_height = <ChainTipHeight<T>>::get();
+
+		match current_height.cmp(&best_height) {
+			Ordering::Equal => {
+				<BestHeaders<T>>::append(hash);
+			}
+			Ordering::Greater => {
+				<BestHeaders<T>>::kill();
+				<BestHeaders<T>>::append(hash);
+				<ChainTipHeight<T>>::put(current_height);
+			}
+			Ordering::Less => {
+				// This is fine. We can still have a valid header, but it might just be on a
+				// different fork and at a lower height than the "best" overall header.
+			}
+		}
+
+		if header.requires_justification {
```

Comment:
Probably think about changing that as a part of #371 - if `requires_justification` is `true` then the entry is inserted into `RequiresJustification` in this function. But it is `false`, then it is not removed here, but instead in a separate `update_best_finalized`. That is not super-intuitive

Reply:
Good point. I'm still struggling a bit with getting the abstractions right for storage. I think this specific concern should be something that can be fixed fairly easily in this PR

Reply:
Actually, I just wanted to note that this struct needs some clearer API :) The way you've changed that adds 2 storage ops (read + write) for (almost) every header that we update. And now we update headers that we're finalizing. So when finalizing 10 headers it would be 10 additional reads + 10 additional writes :)

Let's leave it as is for now, but definitely it is something to think about when working on #371

Reply:
(btw - I think `contains_key` call is redundant - you may just call `remove` in else)